### ISSUE TITLE
Optimization: Cash kernels and move creation to correct device

### ISSUE THREAD

**Issue by zakajd** [photosynthesis-team/piq#227]
**Is your feature request related to a problem? Please describe.**
Most metrics use some kind of kernels for extraction of image features.
Now those kernels are created each time when metric is called. That's slow and redundant when the function is called hundreds of times.

Another problem related to performance is creation of temporal tensors first on CPU and later moving them to GPU if needed (`.to(x.device)`). 

**Describe the solution you'd like**
1. Add support for cashing kernels between runs. Add `kernel` param to functional API. Create and store kernel in class API
2. Replace `.to(x.device)` and simmular calls to explicit creation of tensor on target device. Like `torch.ones(N, N, device=x.device)`.

**Additional context**
For metrics that use more than one kernel implementation details can be discussed.

**Comment by denproc** [comment]
Hi @zakajd,

The second part is partially resolved in #334.

While the proposed changes cover the tensor creation related to filters and colour conversions, some parts of the library could still use `.to` (`piq/base.py`,  `fid.py`, `perceptual.py, `pieapp.py` and `piq/functional/resize.py`).
